Decrease use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF/

cdumez opened the pull request:
#### 5c846c99d1332f366820911b12c069b26a1745a1
<pre>
Decrease use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF/
<a href="https://bugs.webkit.org/show_bug.cgi?id=283035">https://bugs.webkit.org/show_bug.cgi?id=283035</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/persistence/PersistentCoders.cpp:
* Source/WTF/wtf/persistence/PersistentDecoder.cpp:
(WTF::Persistence::Decoder::bufferPointerForDirectRead):
(WTF::Persistence::Decoder::decodeFixedLengthData):
(WTF::Persistence::Decoder::decodeNumber):
* Source/WTF/wtf/persistence/PersistentDecoder.h:
* Source/WTF/wtf/persistence/PersistentEncoder.cpp:
(WTF::Persistence::Encoder::grow):
(WTF::Persistence::Encoder::updateChecksumForData):
(WTF::Persistence::Encoder::encodeFixedLengthData):
(WTF::Persistence::Encoder::encodeNumber):
(WTF::Persistence::Encoder::encodeChecksum):
* Source/WTF/wtf/persistence/PersistentEncoder.h:
(WTF::Persistence::Encoder::updateChecksumForNumber):
* Source/WTF/wtf/text/cf/AtomStringImplCF.cpp:
(WTF::AtomStringImpl::add):
* Source/WTF/wtf/text/cf/StringImplCF.cpp:
* Source/WTF/wtf/text/cocoa/ContextualizedNSString.mm:
(-[WTFContextualizedNSString getCharacters:range:]):
* Source/WebCore/platform/WebCorePersistentCoders.cpp:
(WTF::Persistence::decodeCFData):

Canonical link: <a href="https://commits.webkit.org/286554@main">https://commits.webkit.org/286554@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97f2b4a0a170c2cef3ed597f59ca52d0d978faa7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76320 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29222 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80831 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27592 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78436 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3646 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/59849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/17968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79387 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/49753 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65545 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/40189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47151 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23032 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/25914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69499 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/68278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23366 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82287 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75596 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3692 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2415 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3846 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65517 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67377 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11345 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9444 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97850 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11806 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3640 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6447 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21409 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3663 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7092 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5421 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/cocoa/ContextualizedNSString.mm:
```diff
@@ -67,9 +65,9 @@ - (void)getCharacters:(unichar *)buffer range:(NSRange)range
     auto contentsSubstring = contents.substring(contentsLow - context.length(), contentsHigh - contentsLow);
     static_assert(std::is_same_v<std::make_unsigned_t<unichar>, std::make_unsigned_t<UChar>>);
     contextSubstring.getCharacters(reinterpret_cast<UChar*>(buffer));
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
     contentsSubstring.getCharacters(reinterpret_cast<UChar*>(buffer) + contextSubstring.length());
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
```
I presume we’ll fix this one when we fix `getCharacters` to take a `span` argument?

cdumez replied:
Yes, this is a bigger refactoring so I kept it out of this PR.

darinadler commented on Source/WebCore/platform/WebCorePersistentCoders.cpp:
```diff
@@ -337,11 +337,11 @@ static std::optional<RetainPtr<CFDataRef>> decodeCFData(Decoder& decoder)
         return std::nullopt;
 
-    return adoptCF(CFDataCreate(nullptr, pointer, *size));
+    return adoptCF(CFDataCreate(nullptr, buffer.data(), buffer.size()));
```
Don’t we have a wrapper somewhere to make a `CFData` by copying a `span`?

cdumez replied:
I think we do. I'll update.

cdumez committed:
Decrease use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF/
https://bugs.webkit.org/show_bug.cgi?id=283035

Reviewed by Darin Adler.

* Source/WTF/wtf/persistence/PersistentCoders.cpp:
* Source/WTF/wtf/persistence/PersistentDecoder.cpp:
(WTF::Persistence::Decoder::bufferPointerForDirectRead):
(WTF::Persistence::Decoder::decodeFixedLengthData):
(WTF::Persistence::Decoder::decodeNumber):
* Source/WTF/wtf/persistence/PersistentDecoder.h:
* Source/WTF/wtf/persistence/PersistentEncoder.cpp:
(WTF::Persistence::Encoder::grow):
(WTF::Persistence::Encoder::updateChecksumForData):
(WTF::Persistence::Encoder::encodeFixedLengthData):
(WTF::Persistence::Encoder::encodeNumber):
(WTF::Persistence::Encoder::encodeChecksum):
* Source/WTF/wtf/persistence/PersistentEncoder.h:
(WTF::Persistence::Encoder::updateChecksumForNumber):
* Source/WTF/wtf/text/cf/AtomStringImplCF.cpp:
(WTF::AtomStringImpl::add):
* Source/WTF/wtf/text/cf/StringImplCF.cpp:
* Source/WTF/wtf/text/cocoa/ContextualizedNSString.mm:
(-[WTFContextualizedNSString getCharacters:range:]):
* Source/WebCore/platform/WebCorePersistentCoders.cpp:
(WTF::Persistence::decodeCFData):

Canonical link: https://commits.webkit.org/286554@main
